cloudify_cli/commands/upgrade.py

```diff
@@ -0,0 +1,128 @@
+
+REMOTE_WORKFLOW_STATE_PATH = '/opt/manager/_workflow_state.json'
+
+logger = get_logger()
+
+
+def upgrade(validate_only,
+            skip_validations,
+            blueprint_path,
+            inputs,
+            install_plugins,
+            task_retries,
+            task_retry_interval,
+            task_thread_pool_size):
+
+    management_ip = utils.get_management_server_ip()
+    client = utils.get_rest_client(management_ip)
+
+    curr_status = client.maintenance_mode.status().status
+    if curr_status == MAINTENANCE_MODE_DEACTIVATED:
+        msg = "'Manager must be in maintenance mode for upgrade to run."
+        error = exceptions.CloudifyCliError(msg)
+        error.possible_solutions = [
+            "Activate maintenance mode using 'cfy maintenance-mode activate'"
+        ]
+        raise error
+    elif curr_status == MAINTENANCE_MODE_ACTIVATING:
+        _wait_for_maintenance(client)
+
+    env_name = 'manager-upgrade'
+    # init local workflow execution environment
+    env = common.initialize_blueprint(blueprint_path,
+                                      storage=None,
+                                      install_plugins=install_plugins,
+                                      name=env_name,
+                                      inputs=inputs)
+
+    if not skip_validations:
+        logger.info('Executing upgrade validation')
+        env.execute(workflow='execute_operation',
+                    parameters={'operation':
+                                'cloudify.interfaces.validation.creation'},
+                    task_retries=task_retries,
+                    task_retry_interval=task_retry_interval,
+                    task_thread_pool_size=task_thread_pool_size)
+
+    if not validate_only:
+        with open(inputs) as f:
+            inputs_data = yaml.load(f)
+        fabric_data = {'host_string': management_ip,
+                       'user': inputs_data['ssh_user'],
+                       'key_filename': inputs_data['ssh_key_filename']}
+
+        logger.info('Starting manager upgrade process.')
+        _put_workflow_state_file(fabric_data, is_upgrade=True)
+        try:
+            logger.info('Executing manager upgrade.')
+            env.execute('install',
+                        task_retries=task_retries,
+                        task_retry_interval=task_retry_interval)
+        except Exception as e:
+            logger.info('Upgrade process failed. Starting manager '
+                        'rollback')
+            raise e
+
+    logger.info('Deactivating manager maintenance-mode')
+    client.maintenance_mode.deactivate()
+
+
+def _put_workflow_state_file(fabric_data, is_upgrade):
+    manager_state_file = tempfile.NamedTemporaryFile(delete=True)
+    content = {'is_upgrade': is_upgrade}
+    with open(manager_state_file.name, 'w') as f:
+        f.write(json.dumps(content))
+
+    with _manager_fabric_env(**fabric_data) as api:
+        api.put(use_sudo=True,
```

Comment:
`ssh.py` contains a function called `get_file_from_manager`. Please add another function there called `put_file_in_manager` which implements the same logic so that you don't have to implement it specifically here.
Since it already includes the logic for retrieving everything like host_string, etc.. you won't have to use the fabric data thing above.